### DISCUSSION TITLE
Storage manager handle error, best practice for database

### DISCUSSION THREAD

**Discussion by nguyenvoanhkhoa1** [GrapesJS/grapesjs#5369] [Q&A] [answered]
This is my current `storageManager`:
```typescript
storageManager: {
        type: "remote",
        autosave: false,
        options: {
          remote: {
            urlStore: `${process.env.NEXT_PUBLIC_API_URL}projects/${projectId}`,
            urlLoad: `${process.env.NEXT_PUBLIC_API_URL}projects/${projectId}`,
            fetchOptions: (opts) => (opts.method === "POST" ? { method: "PATCH" } : {}),
            headers: {
              "Content-Type": "application/json",
              Authorization: `Bearer ${token}`,
            },
            onStore: (data) => ({
              data: {
                assets: JSON.stringify(data.assets),
                pages: JSON.stringify(data.pages),
                styles: JSON.stringify(data.styles),
              },
            }),
            onLoad: (result) => {
              console.log(result);
              return {
                assets: JSON.parse(result.data.assets),
                pages: JSON.parse(result.data.pages),
                styles: JSON.parse(result.data.styles),
              };
            },
          },
        },
      }
``` 
My backend code and the setup are working well, but I have 2 questions:
- I use the `JWT` token for authorization, and it has an `expired time`. When The token expires, API returns a `401` status code. Is there any config option for `storageManager` to handle `onLoad` and `onStore` error, which is to use the `refreshToken` to get the `accessToken`, or do I need to custom `onLoad` and `onStore` methods?
- I save the project to the database with key data, and it has 3 key assets, pages, and styles with string type, I am new and want to know the best practice for saving to the database.

Thank you.

**Comment by nguyenvoanhkhoa1** [comment]
First question I fixed it by defining new storage and using Axios to handle the error, so it automatically calls refresh token and retries the action when load or store returns 401, but I still need you guys second question

**Reply by artf** [reply, marked as answer]
I'd suggest storing the project data as it is, without trying to recreate your own structure (in the end it's a simple JSON) like you're doing right now.
[Here](https://grapesjs.com/docs/modules/Storage.html#html-code-with-project-data) you can see how the payload is enriched with HTML but the `data` is stored as it is and loaded simply in this way `onLoad: payload => payload.data,`

**Reply by nguyenvoanhkhoa1** [reply]
Thanks, I can see it. The JSON payload is the same size as the string payload in my devtool. Thank you.
